### ISSUE TITLE
Critical bug with Didn't find class "com.google.firebase.iid.FirebaseInstanceIdService"

### ISSUE THREAD

**Issue by N0lex** [AppsFlyerSDK/appsflyer-android-sdk#7]
Hello!

Similar issue  reported by user @ybereza in this topic: https://github.com/AppsFlyerSDK/AppsFlyerAndroidSDK/issues/5

I am getting the same error only in the very first launch.

**Logs:**

```
11-13 20:21:25.495: I/.mygame(27917): Rejecting re-init on previously-failed class java.lang.Class<com.appsflyer.FirebaseInstanceIdListener>: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/firebase/iid/FirebaseInstanceIdService;
11-13 20:21:25.495: I/.mygame(27917):   at com.appsflyer.internal.k$d com.appsflyer.internal.k.?(android.content.Context) (:105)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(android.content.Context, java.util.Map) (:2210)
11-13 20:21:25.495: I/.mygame(27917):   at java.util.Map com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.internal.j) (:1973)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.AppsFlyerLibCore, com.appsflyer.internal.j) (:51010)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore$a.run() (:3298)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Object java.util.concurrent.Executors$RunnableAdapter.call() (Executors.java:459)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.FutureTask.run() (FutureTask.java:266)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run() (ScheduledThreadPoolExecutor.java:301)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AFExecutor$3$2.run() (:33)
11-13 20:21:25.495: I/.mygame(27917):   at void java.lang.Thread.run() (Thread.java:764)
11-13 20:21:25.495: I/.mygame(27917): Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdService" on path: DexPathList[[zip file "/data/app/com.mygame-TeYKg96_CmVgEdDvsD6tNA==/base.apk"],nativeLibraryDirectories=[/data/app/com.mygame-TeYKg96_CmVgEdDvsD6tNA==/lib/arm64, /data/app/com.mygame-TeYKg96_CmVgEdDvsD6tNA==/base.apk!/lib/arm64-v8a, /system/lib64]]
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:134)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
11-13 20:21:25.495: I/.mygame(27917):   at com.appsflyer.internal.k$d com.appsflyer.internal.k.?(android.content.Context) (:105)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(android.content.Context, java.util.Map) (:2210)
11-13 20:21:25.495: I/.mygame(27917):   at java.util.Map com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.internal.j) (:1973)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.AppsFlyerLibCore, com.appsflyer.internal.j) (:51010)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore$a.run() (:3298)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Object java.util.concurrent.Executors$RunnableAdapter.call() (Executors.java:459)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.FutureTask.run() (FutureTask.java:266)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run() (ScheduledThreadPoolExecutor.java:301)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AFExecutor$3$2.run() (:33)
11-13 20:21:25.495: I/.mygame(27917):   at void java.lang.Thread.run() (Thread.java:764)
11-13 20:21:25.495: I/.mygame(27917): Rejecting re-init on previously-failed class java.lang.Class<com.appsflyer.FirebaseInstanceIdListener>: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/firebase/iid/FirebaseInstanceIdService;
11-13 20:21:25.495: I/.mygame(27917):   at com.appsflyer.internal.k$d com.appsflyer.internal.k.?(android.content.Context) (:105)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(android.content.Context, java.util.Map) (:2210)
11-13 20:21:25.495: I/.mygame(27917):   at java.util.Map com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.internal.j) (:1973)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.AppsFlyerLibCore, com.appsflyer.internal.j) (:51010)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore$a.run() (:3298)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Object java.util.concurrent.Executors$RunnableAdapter.call() (Executors.java:459)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.FutureTask.run() (FutureTask.java:266)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run() (ScheduledThreadPoolExecutor.java:301)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AFExecutor$3$2.run() (:33)
11-13 20:21:25.495: I/.mygame(27917):   at void java.lang.Thread.run() (Thread.java:764)
11-13 20:21:25.495: I/.mygame(27917): Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdService" on path: DexPathList[[zip file "/data/app/com.mygame-TeYKg96_CmVgEdDvsD6tNA==/base.apk"],nativeLibraryDirectories=[/data/app/com.mygame-TeYKg96_CmVgEdDvsD6tNA==/lib/arm64, /data/app/com.mygame-TeYKg96_CmVgEdDvsD6tNA==/base.apk!/lib/arm64-v8a, /system/lib64]]
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:134)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
11-13 20:21:25.495: I/.mygame(27917):   at com.appsflyer.internal.k$d com.appsflyer.internal.k.?(android.content.Context) (:105)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(android.content.Context, java.util.Map) (:2210)
11-13 20:21:25.495: I/.mygame(27917):   at java.util.Map com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.internal.j) (:1973)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore.?(com.appsflyer.AppsFlyerLibCore, com.appsflyer.internal.j) (:51010)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AppsFlyerLibCore$a.run() (:3298)
11-13 20:21:25.495: I/.mygame(27917):   at java.lang.Object java.util.concurrent.Executors$RunnableAdapter.call() (Executors.java:459)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.FutureTask.run() (FutureTask.java:266)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run() (ScheduledThreadPoolExecutor.java:301)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
11-13 20:21:25.495: I/.mygame(27917):   at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
11-13 20:21:25.495: I/.mygame(27917):   at void com.appsflyer.AFExecutor$3$2.run() (:33)
11-13 20:21:25.495: I/.mygame(27917):   at void java.lang.Thread.run() (Thread.java:764)
```

Why do not you fix this problem?

Unity 2019.2.0f
Android SDK 9.0 (API Level 28)
Samsung Galaxy S10 (Android 9)
AppsFlyerUnityPlugin_v4.20.3.unitypackage

**Comment by zjw552048** [comment]
I met the same error stack when integrate AF android SDK,  and fixed it by adding 'firebase-core' or 'firebase-messaging' dependency which determind by you SDK version.

**Comment by acerolla** [comment]
> I met the same error stack when integrate AF android SDK, and fixed it by adding 'firebase-messaging' dependency which determind by you SDK version.

How can I determine SDK version I need?

**Comment by GensaGames** [comment]
Any updates on this?

**Comment by Fudds** [comment]
This error is not allowing me to build my project, newest version of Appsflyer 4.22.1

**Comment by iagormoraes** [comment]
Having the same issue on 5.1.3 even with `com.google.firebase:firebase-messaging:17.1.0`, using version 1.4.4 the error doesn't happen but is too old. Any update of the Appsflyer team?

**Comment by N0lex** [comment]
@iagormoraes The Appsflyer company has very poor support. Hardly someone will help. (

**Comment by af-fess** [comment]
> Having the same issue on 5.1.3 even with `com.google.firebase:firebase-messaging:17.1.0`, using version 1.4.4 the error doesn't happen but is too old. Any update of the Appsflyer team?

hi @iagormoraes , where did you take `5.1.3`, do you talk about Unity?

**Comment by m4r00p** [comment]
Any luck with this issue?

**Comment by af-vs** [comment]
Hi everyone!

**TL;DR**
This should not cause any issues other than printing error log and only happens on Android 9 (API level 28). However, it indicates that you miss a dependency that we use for uninstall measurement feature. If you plan to use it, please read the following guide: https://support.appsflyer.com/hc/en-us/articles/210289286-Uninstall-measurement#android-uninstall

**Background**
There are 2 types of error that you can see, based in the version of AppsFlyer SDK being used:
- java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdService" 
- java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.messaging.FirebaseMessagingService"

`com.google.firebase.iid.FirebaseInstanceIdService` is a [deprecated GCM](https://developers.google.com/cloud-messaging/android/android-migrate-fcm)-related class that our latest SDK versions do not have any reference to. You should not use this class and should migrate to FCM if haven't done so yet (I believe it is unlikely that anyone still uses GCM, but no 100% sure)

`com.google.firebase.messaging.FirebaseMessagingService` is an FCM class that we indeed require for Uninstall Measurement feature and we check for it's presence in the runtime. Lack of `com.google.firebase:firebase-messaging` dependency that contains this class will cause Android (only at API level 28) to throw the reported error (because a class that inherits FirebaseMessagingService is being loaded) but it should not crash your app or anything like that.


Taking all of the above into account, I am closing this. If the error above causes crashes, prevents from building or has any other "blocking" behaviour, please reopen with the relevant details.

Thanks!
